cpp/src/replace/replace.cu

```diff
@@ -13,26 +16,551 @@
+                 "Nulls are in values_to_replace column.");
+
+    std::unique_ptr<column> output;
+    if (input_col.nullable() || replacement_values.nullable()) {
+      output = make_numeric_column(input_col.type(), input_col.size(), UNINITIALIZED, stream, mr);
+    }
+    else
+      output = make_numeric_column(input_col.type(), input_col.size(), UNALLOCATED, stream, mr);
+
+    cudf::mutable_column_view outputView = (*output).mutable_view();
+    cudf::experimental::type_dispatcher(input_col.type(),
+                                        replace_kernel_forwarder { },
+                                        input_col,
+                                        values_to_replace,
+                                        replacement_values,
+                                        outputView,
+                                        stream,
+                                        mr);
+
+    CHECK_STREAM(stream);
+    return output;
+  }
+
+} //end details
+namespace experimental {
+/* --------------------------------------------------------------------------*/
+/**
+ * @brief Replace elements from `input_col` according to the mapping `values_to_replace` to
+ *        `replacement_values`, that is, replace all `values_to_replace[i]` present in `input_col`
+ *        with `replacement_values[i]`.
+ *
+ * @param[in] col gdf_column with the data to be modified
+ * @param[in] values_to_replace gdf_column with the old values to be replaced
+ * @param[in] replacement_values gdf_column with the new values
+ *
+ * @returns output gdf_column with the modified data
+ */
+/* ----------------------------------------------------------------------------*/
+  std::unique_ptr<cudf::column> find_and_replace_all(cudf::column_view const& input_col,
+                                                     cudf::column_view const& values_to_replace,
+                                                     cudf::column_view const& replacement_values,
+                                                     rmm::mr::device_memory_resource* mr){
+    return detail::find_and_replace_all(input_col, values_to_replace, replacement_values, mr, 0);
+  }
+} //end experimental
+} //end cudf
+
+namespace{ //anonymous
+
+template <typename Type>
+__global__
+void replace_nulls_with_scalar(cudf::size_type size,
+                               const Type* __restrict__ in_data,
+                               const cudf::bitmask_type* __restrict__ in_valid,
+                               const Type* __restrict__ replacement,
+                               Type* __restrict__ out_data)
+{
+  int tid = threadIdx.x;
+  int blkid = blockIdx.x;
+  int blksz = blockDim.x;
+  int gridsz = gridDim.x;
+
+  int start = tid + blkid * blksz;
+  int step = blksz * gridsz;
+
+  for (int i=start; i<size; i+=step) {
+    out_data[i] = cudf::bit_is_set(in_valid, i)? in_data[i] : *replacement;
+  }
+}
+
+
+template <typename Type>
+__global__
+void replace_nulls_with_column(cudf::size_type size,
```

Comment:
These custom kernels should be replace entirely by a `thrust::transform_if`. 

https://thrust.github.io/doc/group__transformations_gaaf162576bd1daef87b80ea82bfbf8558.html

The predicate is if the input value is valid, and the transformation is to return the corresponding replacement value.

Reply:
I don't see how it's possible to do this. The predicate in thrust::transform_if operates on the value given, which wouldn't know if it's null or not. If this is possible I would appreciate some guidance in how to accomplish it.

Reply:
In that case, you can make the values given to be a `thrust::counting_iterator` and the predicate to be a device lambda that captures the null mask. Then, define the unary function to take an index rather than a value.

Reply:
Have replaced the custom kernels with thrust::for_each_n invocations.